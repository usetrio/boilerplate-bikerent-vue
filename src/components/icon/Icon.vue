<script lang="ts">
import { defineComponent, type PropType } from 'vue'

export enum IconType {
  REGULAR = 'regular',
  SOLID = 'solid'
}

export type PropIconType = `${IconType}`

export default defineComponent({
  name: 'Icon',
  props: {
    type: {
      type: String as PropType<PropIconType>,
      default: IconType.REGULAR
    },
    size: {
      type: String as PropType<UISizing | UISizingType>,
      default: 'base'
    }
  },
  computed: {
    icon() {
      const [icon] = (this.$slots.default && this.$slots.default()) || []

      if (!icon) {
        return ''
      }

      return (icon.children as String).trim()
    },
    componentClass() {
      return ['icon', `text-${this.size}`]
    },
    computedClass() {
      const classes = [`fa-${this.type}`]

      if (this.icon) {
        classes.push(`fa-${this.icon}`)
      }

      return classes
    }
  }
})
</script>

<template>
  <div :class="componentClass">
    <em :class="computedClass" />
  </div>
</template>

<style lang="scss" scoped>
.icon {
  @apply .inline-block;
  position: relative;
}
</style>
