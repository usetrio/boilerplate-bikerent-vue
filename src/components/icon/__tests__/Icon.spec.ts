import { Icon, IconType } from '@/components/icon'
import { describe, expect, it } from 'vitest'
import { flushPromises, mount } from '@vue/test-utils'

describe('BikeBookmark', () => {
  it('renders correctly', () => {
    const wrapper = mount(Icon, { slots: { default: 'star' } })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('shows empty icon', () => {
    const wrapper = mount(Icon)

    expect(wrapper.find('em').classes().length).equal(1)
  })

  it('shows icon solid type', () => {
    const wrapper = mount(Icon, { props: { type: IconType.SOLID }, slots: { default: 'star' } })

    expect(wrapper.find('em').classes()).contain('fa-solid')
  })

  it('shows different size', () => {
    const wrapper = mount(Icon, { props: { size: 'lg' }, slots: { default: 'star' } })

    expect(wrapper.classes()).contain('text-lg')
  })

  it('changes size', async () => {
    const wrapper = mount(Icon, { slots: { default: 'star' } })

    expect(wrapper.classes()).contain('text-base')

    wrapper.setProps({ size: 'lg' })
    await flushPromises()

    expect(wrapper.classes()).contain('text-lg')
  })
})
