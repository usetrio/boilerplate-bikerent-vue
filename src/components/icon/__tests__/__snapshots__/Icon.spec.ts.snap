// Vitest Snapshot v1

exports[`BikeBookmark > renders correctly 1`] = `"<div class=\\"icon text-base\\" data-v-6ea13f32=\\"\\"><em class=\\"fa-regular fa-star\\" data-v-6ea13f32=\\"\\"></em></div>"`;
